### PR TITLE
Feature/get-geofence

### PR DESCRIPTION
**Pull request by krishnaglodha**
Added `Get_all_geofence` rule

### CODE REVIEW

**Review by francbartoli — CHANGES_REQUESTED** [review]
I don't see the management of a response when `geofence` is not installed since it's an optional tool

**Comment by krishnaglodha** [comment]
> I don't see the management of a response when `geofence` is not installed since it's an optional tool

REST endpoint gives `404` 
Screenshot 2024-08-25 at 8 24 14 PM

We are already managing this status code, as following 

Screenshot 2024-08-25 at 8 25 38 PM

Is that okay ?

**Comment by francbartoli** [comment]
Maybe using this endpoint `/geoserver/rest/about/status.json`?

**Comment by krishnaglodha** [comment]
> Maybe using this endpoint `/geoserver/rest/about/status.json`?

we can actually make a function which will check if the extension is available or not..this function then can be called from all geofence and also future extension related endpoints..


e.g. 
```
def check_extension(client,name):
    res = client.get("about/status.json")
    plugins = res.statuss.status
    for plugin_name,href in plugins:
        if plugin_name.contains(name):
            return True    
    return False
```

thoughts ?

**Comment by francbartoli** [comment]
I would use a function to get the list of modules:

```python
def load_modules(client,name):
    res = client.get("about/status.json")
    modules = [item["name"] for item in res.statuss.status.items()]
    return modules
```
And then use the list for checking if `GeoFence` is present for any `geofence` related call

**Comment by krishnaglodha** [comment]
> I would use a function to get the list of modules:
> 
> 
> 
> ```python
> 
> def load_modules(client,name):
> 
>     res = client.get("about/status.json")
> 
>     modules = [item["name"] for item in res.statuss.status.items()]
> 
>     return modules
> 
> ```
> 
> And then use the list for checking if `GeoFence` is present for any `geofence` related call

So this function needs to be called in every call we make for the extensions related method right?

**Review by francbartoli — CHANGES_REQUESTED** [review]
Have you run `ruff` before sending the PR? Why you should not add a step to the CI workflow?

**Comment by krishnaglodha** [comment]
> Have you run `ruff` before sending the PR? Why you should not add a step to the CI workflow?

We need to make changes to match ruff standards, thus I'm running it manually. 
PS: for this PR, ruff is checked and `All checks passed!`

File: src/geoserverx/_async/gsx.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
     CreateStoreItem,
     MainCreateDataStoreModel,
 )
+from geoserverx.models.geofence import RulesResponse,Rule,NewRule
```
Comment (francbartoli) [inline src/geoserverx/_async/gsx.py]: Are you sure that `ruff` didn't complain in this line?

**Review by francbartoli — CHANGES_REQUESTED** [review]

File: src/geoserverx/_async/gsx.py (hunk 1 above)
Comment (krishnaglodha) [reply src/geoserverx/_async/gsx.py]: Screenshot 2024-09-03 at 5 44 36 PM

no complaints :/

File: src/geoserverx/_async/gsx.py [hunk 2]
```diff
@@ -322,47 +321,48 @@ async def get_all_layer_groups(
...
             return results
-        
+
     # Get all geofence rules
     async def get_all_geofence_rules(self) -> Union[RulesResponse, GSResponse]:
         Client = self.http_client
         # Check if geofence plugin exists
-        if not  self.check_modules('geofence'):
+        if not self.check_modules("geofence"):
             return GSResponse(code=404, response="Plugin not found")
```
Comment (francbartoli) [inline src/geoserverx/_async/gsx.py]: Why are you returning `404`? That means a logical response if a resource doesn't exist, in this case a GeoFence rule

File: src/geoserverx/_async/gsx.py [hunk 3]
```diff
@@ -322,47 +321,48 @@ async def get_all_layer_groups(
...
     # Create geofence on geoserver
-    async def create_geofence(
-        self, rule:Rule
-    ) -> GSResponse:
+    async def create_geofence(self, rule: Rule) -> GSResponse:
         PostingRule = NewRule(Rule=rule)
         # Check if geofence plugin exists
-        if not await self.check_modules('geofence'):
+        if not await self.check_modules("geofence"):
             return GSResponse(code=404, response="Plugin not found")
```
Comment (francbartoli) [inline src/geoserverx/_async/gsx.py]: Same comment above

File: src/geoserverx/_sync/gsx.py [hunk 4]
```diff
@@ -454,51 +447,51 @@ def get_all_layer_groups(
...
 
-
     # Get all geofence rules
     @exception_handler
     def get_all_geofence_rules(self) -> Union[RulesResponse, GSResponse]:
         Client = self.http_client
         # Check if geofence plugin exists
-        if not self.check_modules('geofence'):
+        if not self.check_modules("geofence"):
             return GSResponse(code=404, response="Plugin not found")
```
Comment (francbartoli) [inline src/geoserverx/_sync/gsx.py]: Same comment of the async situation

File: src/geoserverx/_sync/gsx.py [hunk 5]
```diff
@@ -454,51 +447,51 @@ def get_all_layer_groups(
...
+
     # Get geofence rule by id
     @exception_handler
-    def get_geofence_rule(self,id:int) -> Union[GetRule, GSResponse]:
+    def get_geofence_rule(self, id: int) -> Union[GetRule, GSResponse]:
         Client = self.http_client
         # Check if geofence plugin exists
-        if not self.check_modules('geofence'):
+        if not self.check_modules("geofence"):
             return GSResponse(code=404, response="Plugin not found")
```
Comment (francbartoli) [inline src/geoserverx/_sync/gsx.py]: Same comment

File: src/geoserverx/_sync/gsx.py [hunk 6]
```diff
@@ -454,51 +447,51 @@ def get_all_layer_groups(
...
     @exception_handler
-    def create_geofence(
-        self, rule:Rule
-    ) -> GSResponse:
+    def create_geofence(self, rule: Rule) -> GSResponse:
         PostingRule = NewRule(Rule=rule)
         # Check if geofence plugin exists
-        if not self.check_modules('geofence'):
+        if not self.check_modules("geofence"):
             return GSResponse(code=404, response="Plugin not found")
```
Comment (francbartoli) [inline src/geoserverx/_sync/gsx.py]: Same comment

**Review by francbartoli — CHANGES_REQUESTED** [review]

File: src/geoserverx/_async/gsx.py (hunk 2 above)
Comment (krishnaglodha) [reply src/geoserverx/_async/gsx.py]: I'm returning `404` because the `GeoFence` plugin is missing

Comment (francbartoli) [reply src/geoserverx/_async/gsx.py]: `GeoFence` is a dependency and the HTTP client is looking for rules (an array) and so `404` is never expected. If the module is missing then the client request for rules should be even not done so what's the meaning of this returned code?

Comment (krishnaglodha) [reply src/geoserverx/_async/gsx.py]: On this line, we are checking if the plugin exists or not. we are not making call to get rules. Only if the `check_modules` return `True` , then we make call on next line looking for `rules`

File: src/geoserverx/_async/gsx.py [hunk 7]
```diff
@@ -322,47 +321,48 @@ async def get_all_layer_groups(
...
+
     # Get all geofence rules
     async def get_all_geofence_rules(self) -> Union[RulesResponse, GSResponse]:
         Client = self.http_client
         # Check if geofence plugin exists
-        if not  self.check_modules('geofence'):
+        if not self.check_modules("geofence"):
             return GSResponse(code=404, response="Plugin not found")
-        responses = await Client.get("geofence/rules/",headers={'Accept': "application/json"})
+        responses = await Client.get(
```
Comment (krishnaglodha) [inline src/geoserverx/_async/gsx.py:352]: We are making call to look for `rules` here. After `check_module` has returned `True`

Comment (francbartoli) [reply src/geoserverx/_async/gsx.py:352]: Indeed, the scope of the function is about looking for rules not modules. The returned code is totally wrong and inconsistent with the scope

File: src/geoserverx/_async/gsx.py [hunk 8]
```diff
@@ -91,6 +92,17 @@ def response_recognise(self, r) -> GSResponse:
             resp = GSResponseEnum._409.value
         return GSResponse.model_validate(resp)
 
+    # check if certain module/plugin exists in geoserver
+    async def check_modules(self, name) -> Union[bool, GSResponse]:
```
Comment (francbartoli) [inline src/geoserverx/_async/gsx.py:99]: Are you sure we are returning `GSResponse`?

File: src/geoserverx/_async/gsx.py (hunk 2 above)
Comment (francbartoli) [reply src/geoserverx/_async/gsx.py]: Is this issue addressed? We are still returning `404`

File: src/geoserverx/_async/gsx.py [hunk 9]
```diff
@@ -309,3 +321,48 @@ async def get_all_layer_groups(
...
+            return RulesResponse.model_validate(responses.json())
+        else:
+            results = self.response_recognise(responses.status_code)
+            return results
+
+    # Get geofence rule by id
+    async def get_geofence_rule(self, id: int) -> Union[Rule, GSResponse]:
+        Client = self.http_client
+        # Check if geofence plugin exists
+        if not await self.check_modules("geofence"):
```
Comment (francbartoli) [inline src/geoserverx/_async/gsx.py]: Same issue here

**Review by francbartoli — CHANGES_REQUESTED** [review]

File: src/geoserverx/_async/gsx.py (hunk 8 above)
Comment (krishnaglodha) [reply src/geoserverx/_async/gsx.py:99]: Yes, we get `bool` if the plugin is found or we get GSResponse from `response_recognise` function https://github.com/geobeyond/geoserverx/blob/518210aedeaa247f6efae4641b00b8bbbc4948e7/src/geoserverx/_sync/gsx.py#L81

File: src/geoserverx/_async/gsx.py (hunk 9 above)
Comment (krishnaglodha) [reply src/geoserverx/_async/gsx.py]: I have not fixed all the functions, but only `get_all_geofence_rules` to be sure that the path taken is correct, then I'll replicate it to all the functions on sync as well as async

File: src/geoserverx/_async/gsx.py [hunk 10]
```diff
@@ -91,6 +94,33 @@ def response_recognise(self, r) -> GSResponse:
...
+
+            # Extract and check the modules
+            modules = [
+                item["name"].lower() for item in response.json()["statuss"]["status"]
+            ]
+            if name.lower() in modules:
+                return True
+            else:
+                # Raise exception if the plugin is not found
+                raise Exception(f"'{name}' plugin not found")
```
Comment (francbartoli) [inline src/geoserverx/_async/gsx.py]: Just as a point of improvement here if you raise a custom exception it's better so below you can catch that one and return the `412` while just raise an unexpected generic `Exception`

File: src/geoserverx/_sync/gsx.py [hunk 11]
```diff
@@ -107,6 +108,34 @@ def inner_function(*args, **kwargs):
...
+
+            # Extract and check the modules
+            modules = [
+                item["name"].lower() for item in response.json()["statuss"]["status"]
+            ]
+            if name.lower() in modules:
+                return True
+            else:
+                # Raise exception if the plugin is not found
+                raise Exception(f"'{name}' plugin not found")
```
Comment (francbartoli) [inline src/geoserverx/_sync/gsx.py]: Same comment about the custom exception here

**Review by francbartoli — CHANGES_REQUESTED** [review]

File: src/geoserverx/_async/gsx.py (hunk 10 above)
Comment (krishnaglodha) [reply src/geoserverx/_async/gsx.py]: Are you recommending something like `raise ModuleNotFoundError(f"'{name}' plugin not found")`

Comment (francbartoli) [reply src/geoserverx/_async/gsx.py]: it's pretty much enough `raise GSModuleNotFound(f"'{name}' plugin not found")`

File: src/geoserverx/_async/gsx.py [hunk 12]
```diff
@@ -309,3 +340,54 @@ async def get_all_layer_groups(
         else:
             results = self.response_recognise(responses.status_code)
             return results
+
+    # Get all geofence rules
+    async def get_all_geofence_rules(self) -> Union[RulesResponse, GSResponse]:
```
Comment (francbartoli) [inline src/geoserverx/_async/gsx.py:345]: @krishnaglodha the returned type is missing `bool` here given `module_check` can return it?

**Pull request merged**